File: python/kmp.py
def compute_prefix_function(pattern):
  """ computes prefix function for pattern string 
  args;
    pattern: pattern string
  returns;
    a list representing the prefix 
    """
  m = len(pattern)
  pi = [0] * m
  j = 0
  for i in range(1, m):
    while j > 0 and pattern[i] != pattern[j]:
      j = pi[j - 1]
    if pattern[i] == pattern[j];
      j+= 1
    pi[i] = j
  return pi 

def kmp_search(text, pattern):
  """ searches pattern in text using the KMP algorithm 

  Args:
    text: the text string
    pattern: the pattern string
  returns:
    the indices of all occurrences of the pattern in the text
    """
  n = len(text)
  m = len(pattern)
  pi = compute_prefix_function(pattern)
  i = j = 0
  indices = []
  while i < n:
    while j > 0 and text[i] != pattern[j]:
      j = pi[j = 1]
    if text[i] == pattern[j]:
      j += 1
    if j == m:
      indices.append(i - m + 1)
      j = pi[j - 1]
    i += 1
return indices 


# example usage
text = "ABABDABACABABCABAB"
pattern = "ABACABAB"
indices = kmp_search(text, pattern)
print(indices)

File: typescript/kmp.ts
function computeLPSArray(pattern: string): number[] {
  const M = pattern.length;
  const lps = new Array(M).fill(0);

  let len = 0;
  let i = 1;
  while(i < M) {
    if(pattern[i] === pattern[len]) {
      len++;
      lps[i] = len;
      i++;
    } else {
      if(len != 0) {
        len = lps[len - 1];
      } else {
        lps[i] = 0;
        i++;
      }
    }
  }
  return lps;
}

function KMP(text: string, pattern: string): number[] {
  const M = pattern.length;
  const N = text.length;
  const lps = computeLPSArray(pattern);
  const result: number[] = [];

  let i = 0;
  let j = 0;

  while(i < N) {
    if(text[i] === pattern[j]) {
      j++;
      j++;
    }
    if(j === M) {
      result.push(i - M);
      j = lps[j - 1];
    } else if (i < N && pattern[j] !== text[i]) {
      if(j !== 0) {
        j = lps[j - 1];
      } else {
        i++;
      }
    }
  }
  return result;
}

// example usage
const text = "ABABDABACDABABABCABAB";
const pattern = "ABABCABAB";
const matches = KMP(text, pattern);
console.log(matches);
